# Classification  

#### In this example we implement the various Machine Learning Classification Algorithms to perdict the gender using certain data points. We use Scikit learn for this.

First of all import all the dependencies we will be needing to implement the algorithms

In [1]:
#Importing the dependencies
from sklearn import tree
from sklearn import svm
from sklearn.naive_bayes import GaussianNB
from sklearn import neighbors
from sklearn.metrics import accuracy_score

Now we create the classifiers using the sklearn library and store them in a variable.

In [2]:
# Creating the classifiers

# Classifier implementing Decision tree classifier
clfDtree = tree.DecisionTreeClassifier()

# Classifier implementing Support vector machines
clfSVM = svm.SVC()

# Classifier implementing Gaussian Naive Bayes
clfGNB = GaussianNB()

# Classifier implementing the k-nearest neighbors
clfKNN = neighbors.KNeighborsClassifier()

Now we consider a training set on which we want to train our model and predict the results for the future test data.

In [3]:
# Training set

# [height, weight, shoeSize] 
X = [[181, 80, 44], [177, 70, 43], [160, 60, 38], [154, 54, 37], [166, 65, 40], 
     [190, 90, 47], [175, 64, 39], [177, 70, 40], [159, 55, 37], [171, 75, 42], [181, 85, 43]]

# [male or female]
Y = ['male', 'male', 'female', 'female', 'male', 'male', 'female', 'female', 'female', 'male', 'male']

Now we fit our models with our training datasets, in other words we train our models

In [4]:
# Training the models on the training set data

# Decision Tree
clfDtree = clfDtree.fit(X, Y)

# SVM
clfSVM = clfSVM.fit(X, Y)

# Gaussian Naive Bayes
clfGNB = clfGNB.fit(X, Y)

# K-nearest neighbors
clfKNN = clfKNN.fit(X, Y)

Now we consider a test data set and we predict the results using our trained models on the dataset X and Y

In [82]:
# Test set of the data [height, weight, shoeSize]
T = [[160, 50, 38], [157, 47, 40], [190, 70, 43], [184, 69, 52], [160, 38, 40], 
     [150, 30, 27]]

# Test set for the gender
U = ['female', 'male', 'male', 'male', 'female', 'female']

##### Now one by one we predict the output of our test set using our trained models

In [84]:
# Prediction using Decision Tree

predictionDtree = clfDtree.predict(T)
print(predictionDtree)

['female' 'male' 'male' 'male' 'male' 'female']


In [85]:
# Prediction using SVM

predictionSVM = clfSVM.predict(T)
print(predictionSVM)

['male' 'male' 'male' 'male' 'male' 'male']


In [86]:
# Prediction using GNB

predictionGNB = clfGNB.predict(T)
print(predictionGNB)

['female' 'female' 'male' 'male' 'female' 'male']


In [87]:
# Prediction using KNN

predictionKNN = clfKNN.predict(T)
print(predictionKNN)

['female' 'female' 'male' 'male' 'female' 'female']


#### Lets compare the predictions with the expected results(U), and see which model is more accurate

In [88]:
# Comparing the results 

print('Accuracy score for Decision Tree = ',accuracy_score(U, predictionDtree))

Accuracy score for Decision Tree =  0.833333333333


In [89]:
print('Accuracy score for SVM = ',accuracy_score(U, predictionSVM))

Accuracy score for SVM =  0.5


In [90]:
print('Accuracy score for Gaussian Naive Bayes = ',accuracy_score(U, predictionGNB))

Accuracy score for Gaussian Naive Bayes =  0.666666666667


In [91]:
print('Accuracy score for K-nearest neighbors = ',accuracy_score(U, predictionKNN))

Accuracy score for K-nearest neighbors =  0.833333333333


We can see from the above example that Decision tree and K-nearest neighbors did a very good job while predicting the final results as compared to the other models, their scores are equal i.e. 0.834; Gausian naive Bayes stood second last and SVM stood in the last with the least score of 0.5.

<b>DTree(0.834) = KNN(0.834) > GNB(0.67) > SVM(0.5)</b>